### PR TITLE
Implement Dhall Haskell's semi-semantic import caching

### PR DESCRIPTION
**Pull request by TimWSpence**
- Significant improvements in import resolution performance (we do a _lot_ less type checking now)

- Allows us to enable the type-checking prelude test

### CODE REVIEW

**Comment by travisbrown** [comment]
Nice, thanks! Taking a look now.

File: modules/imports/src/test/scala/org/dhallj/imports/ImportResolutionSuite.scala [hunk 1]
```diff
@@ -148,9 +148,9 @@ class ImportResolutionSuite extends FunSuite {
   test("Valid hash") {
 
     val expr = parse(
-      "let x = classpath:/hashed/package.dhall sha256:d60d8415e36e86dae7f42933d3b0c4fe3ca238f057fba206c7e9fbf5d784fe15 in x"
+      "classpath:/hashed/package.dhall sha256:d60d8415e36e86dae7f42933d3b0c4fe3ca238f057fba206c7e9fbf5d784fe15"
     )
-    val expected = parse("let x = 1 in x").normalize
+    val expected = parse("let x = 1 in x").alphaNormalize.normalize
```
Comment (travisbrown) [inline modules/imports/src/test/scala/org/dhallj/imports/ImportResolutionSuite.scala:153]: What's up here? It looks a little odd to alpha-normalize before beta-normalizing.

File: modules/imports/src/main/scala/org/dhallj/imports/ResolveImportsVisitor.scala [hunk 2]
```diff
@@ -175,28 +73,133 @@ private[dhallj] case class ResolveImportsVisitor[F[_] <: AnyRef](cache: ImportsC
...
+                _ <- semanticCache.put(hash, bs)
+              } yield e
+          }
+        } yield e
+
+    def fetch(imp: ImportContext): F[String] = imp match {
+      case Env(value) =>
+        for {
+          vO <- F.delay(sys.env.get(value))
+          v <- vO.fold(
```
[the remark this answers is not in the document]
Comment (TimWSpence) [reply modules/imports/src/main/scala/org/dhallj/imports/ResolveImportsVisitor.scala:113]: Oh nice! I've never noticed that method before. Yes, that should definitely be in a bunch of places

File: modules/imports/src/test/scala/org/dhallj/imports/ImportResolutionSuite.scala (hunk 1 above)
Comment (TimWSpence) [reply modules/imports/src/test/scala/org/dhallj/imports/ImportResolutionSuite.scala:153]: Is it? In my head alpha-normalization belongs at the same stage as computing De Bruijn indices (although obviously not necessary in the same way as Dhall variables are unambiguous)

**Pull request merged**